### ISSUE TITLE
Authentication refactoring and MFA capabilities: TOTP + Lookup Secrets

### ISSUE THREAD

**Issue by aeneasr** [ory/kratos#26]
This is an umbrella issue for all things related to single and second factor authentication, the different factors, passwordless authentication, and more.

* [x] TOTP
* [x] Recovery Codes
* [x] WebAuthn
* [ ] Documentation
* [x] UI examples

Implementation of these features by the ORY Kratos maintainers has officially started: https://github.com/ory/kratos/issues/26#issuecomment-745464249

**Comment by aeneasr** [comment]
The 2FA hook should require 2FA by redirecting to the 2FA endpoint. It should be configurable to cover use cases:

- All users with 2FA enabled should be required to use 2FA
- Support `acr` value from OpenID Connect
- Allow for more complex rules like: *Administrators must always authenticate with 2FA*

> If 2FA is not enabled but required, should users be required to set it up after login or should they be locked out of their accounts?

Enforcement policy:

* Allow grace period before enforcement kicks in
* Allow grace period after sign up to set up 2fa

*Configuration MVP*
```yaml
enable: true/false
policy:
  enforce: true/false
  grace_signup:
  enforce_from: 1.1.2015
```

**Comment by asaf** [comment]
@aeneasr some thoughts about the implementation of this:

- Add a new strategy per 2FA method (e.g., TOTP)
- Flow object should contain other Method(s) for 2FA, this is mandatory for UI to render relevant forms such as OTP input field (maybe the method could contain a hint that this is a stage 2 so UI can render the correct method per stage)
- a login strategy invokes `PostLoginHook` at the end of the flow (if it's not too late),
such 2FA login hook could intercept the request, identify that the user requires 2FA and invoke `UpdateLoginFlowMethod` accordingly. (I wonder if flow needs to be aware of its stage (1FA/2FA) or not, otherwise how UI will know which stage to render?) and redirect back to self-service UI login page
- the UI at this point should render the 2FA methods
- UI submits form where `flow.methods.totp.config.action` should be the URI of the 2FA strategy
- strategy performs authentication and completes the flow (If 2FA is interpreted by `PostLoginHooks`, then it will be invoked twice (once after 1FA, then after 2FA), this time the 2FA hook should be aware of the flow stage in order to avoid an infinite loop and complete it)

**Comment by aeneasr** [comment]
Thank you for these thoughts! Extending on your comment:

1. I think 2fa should be a first class citizen of the login, registration, recovery and settings flow
2. We implement a 2fa flow method (e.g. TOTP) which implements the two required interfaces
3. We either require the login UI to render the 2fa input box, or we add an additional UI URL

If we make 2fa a first-class citizen of the existing flows, I think it will be easier to integrate!

**Comment by sycured** [comment]
Have you an ETA?
It's a mandatory requirement on a project with EHR (Electronic Health Records).

**Comment by aeneasr** [comment]
We don't give out ETAs but if you have a critical dependency you could:

1. Discuss options for contributing 2FA flows (warning: it is going to be a lot of work if you are new to the project)
2. Talk to [jared](jared@ory.sh) about ideas for funding and prioritizing this feature

**Comment by asaf** [comment]
@aeneasr yea, hooking 2FA into other flows but login makes sense, it could be nice to create some generic mechanism to 'suspend' the various flows if 2FA is enabled without repeating this tedious effort.

**Comment by mrosales** [comment]
Just pitching in with some ideas here – I'm not an active user of Kratos (yet!), but I've been following the development and MFA support would be one of the things that helps make the case for migrating much easier.

Including all of these would be way beyond the scope for any kind of initial MFA support, but some ideas that I have that might be helpful for the architecture design are:

1. Ideally MFA allows more than two factors in a single auth flow. An example of the value here could be considering CAPTCHA (which was another tracked ticket) as a requirement of an auth flow in addition to the standard login + second factor. Also in this category of nice to haves would be the ability to replace a password with multiple additional factors. For example email/SMS could be used for step-up 2FA with an identity based on WebAuthn. This would be more of an MFA-first type approach where a password would be considered as just another possible factor in a login rather than an essential part.
2. Different kinds of MFA challenges could likely be grouped into multiple buckets. 
    a. Some, like TOTP can be provided in channel at any point during the flow.
    b. Others, like SMS or an email containing a code might have an out of channel delivery, but in-channel submission (code returned by entering the code or clicking on a redirect link that adds the code to the current session).
   c. Others like a Push Notification (like the Duo or Authy App) would be even more complicated and would require out of channel delivery and status checking. This is the most complicated one because the client logging in won't necessarily know if the request has been satisfied and when it can continue.
3. Ideally the architecture (even if the initial implementation doesn't) would support dynamically defining rules to configure the expected or desired challenges. For example, logging in from a whitelisted IP range could skip MFA for known clients. Some examples of this would be the [Auth0 Rules API](https://auth0.com/docs/rules) or the [AWS Cognito "Define Auth Challenge"](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-define-auth-challenge.html) hook.
4. Another nice future feature could be implementing SRP to do a secure password exchange, and that would likely benefit from the same kind of challenge-response auth flow features that would be required for some of the points in 1-3. While I find the API pretty clunky, AWS Cognito kind of supports this: https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-authentication-flow.html#Using-SRP-password-verification-in-custom-authentication-flow
5. The story for adding/removing MFA devices is pretty complicated. What kind of authentication is required to replace an MFA device? In a 2fa system, maybe the password is enough to change the phone number, but ideally the the account reset or security authorization factor isn't one of the primary factors used for authorization. For a generic MFA interface, I think that this means that its always best to have at least three factors, any two of which can be used to initiate a session and the third can be used to provide an additional factor for performing a reset on either of the other two.

**Comment by znerol** [comment]
> ... MFA support would be one of the things that helps make the case for migrating much easier.

Agree.

> This would be more of an MFA-first type approach where a password would be considered as just another possible factor in a login rather than an essential part.

That is imho a very important observation. In order to bring this issue forward I propose to start defining actionable steps. Starting with key data structures and interfaces:

1. The `Session` member [AuthenticatedAt](https://github.com/ory/kratos/blob/347ad0560126cee4777280fe1c91633d02ce62b8/session/session.go#L25) needs to be replaced with a set of _authentications_. Each _authentication_ record represents a successful authentication attempt at a specific time with a certain _authenticator_. An _authenticator_ represents a reference to the _authentication method_ (credential type) together with a reference to the _credentials_ used (credential id).
1. [NewActiveSession](https://github.com/ory/kratos/blob/master/session/session.go#L47) needs to be extended to take one `authentication` record (instead of an `authenticatedAt` param).
1. A new method needs to be added to the `Session` to `StepUp` a session using an additional authenticator. The `StepUp` method takes an `authentication` record and returns a copy of the session with the `authentication` record added to the existing `authentications` and a [regenerated session id](https://cheatsheetseries.owasp.org/cheatsheets/Session_Management_Cheat_Sheet.html#renew-the-session-id-after-any-privilege-level-change).
1. Similar to the existing `refresh` mechanism (which forces a user to reauthenticate in order to access privileged settings), let's implement a `stepup` mechanism (which forces a user to authenticate with an additional authenticator).

**Comment by aeneasr** [comment]
Thank you for your ideas!

> Ideally MFA allows more than two factors in a single auth flow. An example of the value here could be considering CAPTCHA (which was another tracked ticket) as a requirement of an auth flow in addition to the standard login + second factor. Also in this category of nice to haves would be the ability to replace a password with multiple additional factors. For example email/SMS could be used for step-up 2FA with an identity based on WebAuthn. This would be more of an MFA-first type approach where a password would be considered as just another possible factor in a login rather than an essential part.

I generally agree with this sentiment. It's not very common to have more than one factor but it could be desireable in very specific scenarios. We generally plan to support more than 2FA.

> a. Some, like TOTP can be provided in channel at any point during the flow.

Yup, I think TOTP is the most straight forward to implement.

> b. Others, like SMS or an email containing a code might have an out of channel delivery, but in-channel submission (code returned by entering the code or clicking on a redirect link that adds the code to the current session).

SMS or email auth is what we call "passwordless". I think we need to differentiate between a second factor and the primary authentication mechanism for an identity. It could re-use some code, but I would weight them as conceptually different.

In general, SMS-based authentication is considered a very, very weak mechanism as it is supsectible to so, so many attacks. It is only useful to prevent broad abuse of a system but it is very bad at preventing spear-fishing or other forms of targeted attacks.

> c. Others like a Push Notification (like the Duo or Authy App) would be even more complicated and would require out of channel delivery and status checking. This is the most complicated one because the client logging in won't necessarily know if the request has been satisfied and when it can continue.

The complexity here is probably the user interface as this is what people have to implement themselves. Adding a check here could be a bit cumbersome during implementation but I fear there's no real way around that.

> Ideally the architecture (even if the initial implementation doesn't) would support dynamically defining rules to configure the expected or desired challenges. For example, logging in from a whitelisted IP range could skip MFA for known clients. Some examples of this would be the Auth0 Rules API or the AWS Cognito "Define Auth Challenge" hook.

I think this is a nice extension but I would not consider this as part of the first implementation. In general, I believe this could be implemented using our hooks and some custom logic. It would be very similar to Auth0 rules but it could work in different parts of the pipeline as well, not just immediately after authentication as is the case with Auth0.

> Another nice future feature could be implementing SRP to do a secure password exchange

Never heard of SRP - could you go into a bit of details?

> The story for adding/removing MFA devices is pretty complicated. What kind of authentication is required to replace an MFA device? In a 2fa system, maybe the password is enough to change the phone number, but ideally the the account reset or security authorization factor isn't one of the primary factors used for authorization. For a generic MFA interface, I think that this means that its always best to have at least three factors, any two of which can be used to initiate a session and the third can be used to provide an additional factor for performing a reset on either of the other two.

So how this typically works is that you associate a session with an authentication method (in OIDC this is called ACR). The level of authentication represents the factors you completed. Let's say level 0 for username + password, level 1 for username + password + totp.

When changing TOTP device you would expect the session to have at least level 1 and you could require the user to reconfirm that they are who they are by requesting another level 0 authentication. That's how it is usually done!

You could however also request another level 1 authentication which would mean username + password + existing TOTP.

> The Session member AuthenticatedAt needs to be replaced with a set of authentications. Each authentication record represents a successful authentication attempt at a specific time with a certain authenticator. An authenticator represents a reference to the authentication method (credential type) together with a reference to the credentials used (credential id).

Cool ideas! Another possibility would be to associate a ACR ("Authentication Context Class Reference") level alongside the session which would make it a bit easier to manage. I would actually encourage some reading of the [NIST Digital Identity Guideliens](https://pages.nist.gov/800-63-3/) before determining terminology and processes as they go very deep into details wrt MFA.

**Comment by znerol** [comment]
> Another possibility would be to associate a ACR ("Authentication Context Class Reference") level alongside the session which would make it a bit easier to manage.

Yes, for most use cases that would probably be enough. The [periodic reauthentication scenario](https://pages.nist.gov/800-63-3/sp800-63b.html#423-reauthentication), however, might be difficult to implement unless authentication timestamps are recorded separately for each credential/authenticator pair.

Also linking sessions to credentials in a many-to-many relationship allows to easily kill sessions when credentials are changed (no matter whether those credentials are passwords, fido keys or your fingerprints).

**Comment by zepatrik** [comment]
Another MFA idea we could quite easily implement:

My bank requires

1. password
2. a partial PIN

This partial PIN is a 6 digit PIN, but I only have to input two out of six digits on every login. These two digits are not typed through the keyboard but instead the web page displays a soft keyboard you have to use. I believe the goal is to prevent key loggers and shoulder surfing, as you would have to observe the victim at least a view times to figure out all the digits and their position. Also you have to monitor two channels (keyboard and screen). Quite simple but still effective.

**Comment by aeneasr** [comment]
> Also linking sessions to credentials in a many-to-many relationship allows to easily kill sessions when credentials are changed (no matter whether those credentials are passwords, fido keys or your fingerprints).

I think that makes a lot of sense. However, we probably don't need linkage. We just kill all sessions with auth_time before the change?

> Yes, for most use cases that would probably be enough. The periodic reauthentication scenario, however, might be difficult to implement unless authentication timestamps are recorded separately for each credential/authenticator pair.

Right, I mean it makes sense to add that context to the session!

**Comment by vincentserpoul** [comment]
webauthn can also be used for passwordless, as it can include (depending of the config) 2 factors (what you know and what you have).

**Comment by aeneasr** [comment]
@vincentserpoul but that means that you have one key in the browser probably (software key) and one hardware key, right?

**Comment by znerol** [comment]
@aeneasr exactly. The crucial thing is that you need to *check* whether a given webauthn device is actually a **platform authenticator**.

**Comment by vincentserpoul** [comment]
actually, there is no browser key. If you use a platform authenticator, as @znerol  mentioned, it will leverage a hardware key (the first factor) and to unlock this key, you will need to put a password (something you know) or a biometric (something you are) which is the second factor.

**Comment by aeneasr** [comment]
Oh, I see - so we would actually need two variants - one for a regular login method, and one for a 2fa method?

**Comment by vincentserpoul** [comment]
I think it would make sense! maybe actually all login method could be eligible for 2fa?

**Comment by aeneasr** [comment]
Yes, all login methods should be eligible for 2fa! Once exception is oidc where the oidc provider has 2fa built into their platform (using ACR).

**Comment by vincentserpoul** [comment]
taking a bit of a step back on this, as I was looking into it.
I think MFA would be a better way to think about this functionality. When I read 2fa, I think **second** factor whereas it is actually **two** factor.
Looking at the wikipedia definition: https://en.wikipedia.org/wiki/Multi-factor_authentication
the different factors are: knowledge, possession, inherent, location.

So in the end, the different methods should check one or more of these 4 boxes (password is only knowledge, webauthn is knowledge + possession/location).
two factor would mean: the combination of all methods required to login/register/... should validate at least 2 factors.

Let me know what you think, maybe I'm going to far :).

**Comment by aeneasr** [comment]
You're right with this. I have started working on this today but it is very difficult from a flow perspective. There are several moving parts:

1. We will have multiple authentication methods:
  1. TOTP (this is Google Authenticator)
  2. WebAuthN
  3. SMS
2. For all authentication methods there will be the need to have recovery or backup codes which can be used when the TOTP device, the SMS device, or the device carrying the WebAuthN private key are lost.

Some providers like Coinbase have an elaborate 3FA process where the first step is username + password, the second is SMS or TOTP code, and the third is a link sent to your email confirming the login.

Other providers use adaptive 2FA where 2FA is only required for certain actions or in certain scenarios (e.g. login from unknown device).

It is also possible that there is a combination of several factors (TOTP + WebAuthN) although I have not really seen that in the wild yet.

One primary problem is definitely the onboarding process as it is quite UI heavy. Taken from the TOTP project:

> 1. Provide the user with "recovery codes".
> 1. Generate new TOTP Key for a User. `key,_ := totp.Generate(...)`.
> 1. Display the Key's Secret and QR-Code for the User. `key.Secret()` and `key.Image(...)`.
> 1. Test that the user can successfully use their TOTP. `totp.Validate(...)`.
> 1. Store TOTP Secret for the User in your backend. `key.Secret()`

So there is a multi-step process to "linking" a TOTP device. A similar thing is probably required for SMS and maybe WebAuthN although the last part could be a bit easier.

Another issue is "forced 2FA" or "forced 2FA with enrolment grace period" where users MUST have 2FA set up in order to log in. The problem here is that this would enforce 2FA set up during registration, which is tricky, because the question comes to mind if a user can even register without setting up 2FA or if 2FA is simply disabled and required to set up after successful login.

There are more use cases and flows along these lines I can think of and I think it will be difficult to cover those in a easy to use plug-and-play way as we do currently for login.

This will need some more thinking on my end - we probably have to define the MFA login, recovery, set up flows first and then start looking at the actual implementations of those flows.

**Comment by xlanor** [comment]
For SMS, perhaps I can chime in a little here as we are building our addon microservice to fufill the requirements needed by our business team.

What we've implemented is basically also using TOTP to generate an OTP which is then sent on to the mobile phone number as a text message. When the user enters the OTP, we then consider that as a validated secret and save the secret for future usage.

In the event that the user loses access to the mobile number, there's several ways that this can be worked around - for example, in a similar fashion to downloading generated secrets from aws, for example, the first time that the user logs in after validating the TOTP secret, the keys secret will be available to him for download for the next five minutes.

**Comment by vincentserpoul** [comment]
to clarify things, I think we should make sure what we mean by factor and auth method.
Is **factor** what [wikipedia](https://en.wikipedia.org/wiki/Multi-factor_authentication) refers to: knowledge, possession, inherent, location?
Or is **factor** an authentication method?

I'd lean towards number one, what do you think @aeneasr ?

@xlanor : does it actually make TOTP less secure by sending it through SMS?

**Comment by xlanor** [comment]
@vincentserpoul Yep, I feel that it does, but those were the requirements given to me when I had to implement it due to initial market research suggesting that our user base may not be savvy enough to utilise authenticator tools.

Regarding the **factor** mentioned above, I would agree that factor should be defined as the 4 mentioned in your first.

**Comment by aeneasr** [comment]
Everything below is an excerpt from the NIST Digital Identity Guidelines. While some of those recommendations are for Goverment Agencies only and might be too strict for a freemium mobile game, it does make sense to take their terminology and findings into consideration.

## Authentication and Factors

> The classic paradigm for authentication systems identifies three factors as the cornerstones of authentication:

> - Something you know (e.g., a password).
> - Something you have (e.g., an ID badge or a cryptographic key).
> - Something you are (e.g., a fingerprint or other biometric data).
>
> MFA refers to the use of more than one of the above factors.  The strength of authentication systems is largely determined by the number of factors incorporated by the system — the more factors employed, the more robust the authentication system. For the purposes of these guidelines, using two factors is adequate to meet the highest security requirements. As discussed in Section 5.1, other types of information, such as location data or device identity, may be used by an RP or verifier to evaluate the risk in a claimed identity, but they are not considered authentication factors.

Source: https://pages.nist.gov/800-63-3/sp800-63-3.html#43-authentication-and-lifecycle-management

## Identity Assurance Levels (IAL) / Authenticator Assurance Level (AAL)

<img width="1128" alt="Bildschirmfoto 2020-12-16 um 10 42 05" src="https://user-images.githubusercontent.com/3372410/102331619-5d1b3400-3f8b-11eb-8437-ba1a8933135f.png">

In OpenID Connect, the AAL is used in the ACR (Authentication Context Class Reference). Basically, OpenID ACR could be e.g. `phr` for phishing-resistant and `phrh` for `Phishing-Resistant Hardware-Protected` ([Source](https://openid.net/specs/openid-connect-eap-acr-values-1_0.html))

Given the graphic above, we currently support AAL1 using a Memorized secret and out-of-band (OIDC). We could support OTP (TOTP, SMS) and Crypto (WebAuth) for AAL1 also. So basically adding a new login and registration method called TOTP and WebAuth (I guess this is FIDO2?).

### Out of Band Devices

> An out-of-band authenticator is a physical device that is uniquely addressable and can communicate securely with the verifier over a distinct communications channel, referred to as the secondary channel. The device is possessed and controlled by the claimant and supports private communication over this secondary channel, separate from the primary channel for e-authentication. An out-of-band authenticator is something you have.

#### OTP Devices

- **A single-factor OTP device generates OTPs.** This category includes hardware devices and software-based OTP generators installed on devices such as mobile phones. These devices have an embedded secret that is used as the seed for generation of OTPs and does not require activation through a second factor. The OTP is displayed on the device and manually input for transmission to the verifier, thereby proving possession and control of the device. An OTP device may, for example, display 6 characters at a time. A single-factor OTP device is something you have.Single-factor OTP devices are similar to look-up secret authenticators with the exception that the secrets are cryptographically and independently generated by the authenticator and verifier and compared by the verifier. The secret is computed based on a nonce that may be time-based or from a counter on the authenticator and verifier.
- **A multi-factor OTP device** generates OTPs for use in authentication after activation through an additional authentication factor. This includes hardware devices and software-based OTP generators installed on devices such as mobile phones. The second factor of authentication may be achieved through some kind of integral entry pad, an integral biometric (e.g., fingerprint) reader, or a direct computer interface (e.g., USB port). The OTP is displayed on the device and manually input for transmission to the verifier. For example, an OTP device may display 6 characters at a time, thereby proving possession and control of the device. The multi-factor OTP device is something you have, and it SHALL be activated by either something you know or something you are.

##### Google Authenticator

The Google Authenticator is a single-factor OTP device. It is not a multi-factor OTP device because:

> Multi-factor OTP authenticators operate in a similar manner to single-factor OTP authenticators (see Section 5.1.4.1), except that they require the entry of either a memorized secret or the use of a biometric to obtain the OTP from the authenticator. Each use of the authenticator SHALL require the input of the additional factor.

##### WebAuth

I have not looked into WebAuth too much but I think we can have both single and multi-factor OTP using WebAuth.

#### SMS

Yes SMS is not as secure as the TOTP spec but it is more secure than no second factor and makes broad attacks very hard to pull of. To protect against targeted attacks, TOTP or WebAuth are better suited (e.g. high value targets, goverment agencies, ...).

In the Digital Identity Guidelines, SMS is listed as an out of band authenticator with the limitation that it can only be used in RESTRICTED contexts - I assume RESTRICTED refers to the classification level.

#### Email

> [Authentication] methods that do not prove possession of a specific device, such as voice-over-IP (VOIP) or email, SHALL NOT be used for out-of-band authentication.

https://pages.nist.gov/800-63-FAQ/#q-b11

### Cryptographic Devices

> - **A single-factor software cryptographic authenticator** is a cryptographic key stored on disk or some other "soft" media. Authentication is accomplished by proving possession and control of the key. The authenticator output is highly dependent on the specific cryptographic protocol, but it is generally some type of signed message. The single-factor software cryptographic authenticator is something you have.
> - **A single-factor cryptographic device** is a hardware device that performs cryptographic operations using protected cryptographic key(s) and provides the authenticator output via direct connection to the user endpoint. The device uses embedded symmetric or asymmetric cryptographic keys, and does not require activation through a second factor of authentication. Authentication is accomplished by proving possession of the device via the authentication protocol. The authenticator output is provided by direct connection to the user endpoint and is highly dependent on the specific cryptographic device and protocol, but it is typically some type of signed message. A single-factor cryptographic device is something you have.
> - **A multi-factor software cryptographic authenticator** is a cryptographic key stored on disk or some other "soft" media that requires activation through a second factor of authentication. Authentication is accomplished by proving possession and control of the key. The authenticator output is highly dependent on the specific cryptographic protocol, but it is generally some type of signed message. The multi-factor software cryptographic authenticator is something you have, and it SHALL be activated by either something you know or something you are.
> - **A multi-factor cryptographic device** is a hardware device that performs cryptographic operations using one or more protected cryptographic keys and requires activation through a second authentication factor. Authentication is accomplished by proving possession of the device and control of the key. The authenticator output is provided by direct connection to the user endpoint and is highly dependent on the specific cryptographic device and protocol, but it is typically some type of signed message. The multi-factor cryptographic device is something you have, and it SHALL be activated by either something you know or something you are.

All of these can be implemented using WebAuth iirc.

### Biometrics

> NIST SP 800-63B supports the use of biometrics as an authentication factor with some limitations. Biometrics may be used only as part of multi-factor authentication in conjunction with a specific physical authenticator (something you have). If biometrics are used in multi-factor authentication the following requirements apply:
>
> An authenticated protected channel between sensor and the verifier must be established, or the sensor must be integrated with the endpoint in a manner that resists sensor replacement
The sensor or endpoint must be authenticated prior to capturing the biometric sample from the claimant
The biometric system must operate with a False Match Rate (FMR) of 1 in 1000 or better [see ISO/IEC 2382-37]. This FMR applies under conditions of a conformant attack (i.e., zero-effort impostor attempt) as defined in ISO/IEC 30107-1
The biometric system must limit consecutive failed authentication attempts
The verifier must make a determination of sensor and endpoint performance, integrity, and authenticity based on the requirements presented above.
> The unlocking of a device through biometric match may not be considered to meet these requirements since it is generally not possible for the verifier to obtain any information on how, or whether, the device was unlocked.
>
> NIST recommends that the biometric system implement Presentation Attack Detection (PAD). NIST is considering PAD implementation as a requirement in the future.
>
> Please see SP 800-63B Section 5.2.3 for additional implementation guidance for the use of biometrics as an authentication factor.
>
> Most authenticators using biometrics will function as multi-factor authenticators where the biometric unlocks a secret stored in the authenticator. A memorized secret (often a PIN) can also be used to unlock the same secret, as is common in many mobile devices.

**Comment by aeneasr** [comment]
So I looked a bit into federation and MFA. As far as I can tell, none of the big platforms support any sort of ACR - meaning that you can not ask to "Sign in with Google (and enforce 2fa)". The "(enforce 2fa)" portion is probably linked to your account settings and can apparently not be modified from a 3rd party. Additionally, the ID Token [does not provide any information about the factors used](https://developers.google.com/identity/protocols/oauth2/openid-connect#an-id-tokens-payload).

It is the same story for GitHub and probably a lot of other platforms as well. It seems like ACR is something targeted at bespoke solutions.

For us this means that MFA needs to be enforced in ORY Kratos. Even if a user signs in with Google and presents a valid second factor to Google, we would still need to request the second factor for that user in ORY Kratos, because we have no way to tell if the login used 2fa or not.

This finding is not terrible, because it makes the flow easier to model, as every flow starts with identifying the user (e.g. username+password, google ID token) and then proceeds to the MFA processing.

**Comment by aeneasr** [comment]
I looked more into WebAuthN (FIDO2) and am pretty disappointed. There are several hard obstacles when using WebAuthN without a FIDO2 hardware key (e.g. YubiKey). So these limitations apply to the TPM authenticator type (FaceID, TouchID, ...).

First, it is not possible to port the private key performing the proof authentication across **browsers nor devices** using the WebAuthN standard (https://github.com/w3c/webauthn/issues/151 https://github.com/w3c/webauthn/issues/865). Trying to register in Safari and then logging in on Chrome on the same laptop using WebAuthN will result in a failure.

Apparently, it is not wished for this to be possible either - because of attestation:

> The position Yubico has taken for our Security Key and YubiKey authenticators is that private keys can never leave the authenticator. One reason for this is the issue of attestation: the attestation statement sent in a registration request attests that the private key is created and owned by the authenticator, and has never been exposed to any other party. If it were possible to import a private key from outside the authenticator, we would not be able to attest this. - [Source](https://github.com/w3c/webauthn/issues/865#issuecomment-380434642)

Unfortunately, this means that it is not possible to use the same private key on multiple platforms. Further, this implies that using WebAuthN as the only authentication method is unfeasable because device loss leads to a complete lock out. Without account recovery mechanisms such as security codes, it is not possible to link other browsers.

In total, what this means, is that WebAuthN in TPM mode can not be used as a first party authentication mechanism for web applications because it is bound to exactly one device. Forcing the user to go through several account recovery processes in order to link additional devices or browsers (private keys) is a user experience nightmare. One possibility could be to use the linked TPM hardware to authenticate a login request from another device, but this already sounds very difficult to implement and explain to users - what happens if you first sign up on a laptop?

If one or more YubiKeys are used, you can plug the key into the device. This means that you could theoretically authenticate in multiple browsers, by having the YubiKey in your USB slot. However, almost noone has a YubiKey for personal use - the average consumer owns a smartphone and a laptop or tablet.

All of this makes FIDO2/WebAuthN much less attractive for broad implementation. However, allowing hardware keys in MFA flows does have it's appeal to high-security environments. But having WebAuthN enabled MFA does not remove the need for a back up mechanism (like security codes) which help recover access to an account.

**Comment by vincentserpoul** [comment]
I'm pretty familiar with webauthn.
I totally agree that registering with only webauthn is not possible, for the inconveniences you mentioned.

That said, it does not forbid you to sign in/auth with webauthn.
The most user friendly flow I 've seen:
* register/recovery with magic link in email
* first register, associate device (TPM) with account
* next logins: either use a magic link again, or simply use webauthn.

I think it's really a nice experience for the customer, it makes things really easy.

**Comment by aeneasr** [comment]
Thank you for this use case - I think it makes sense if WebAuthN is seen as an ease-of-use utility versus an actual, strong authentication mechanism. It does make sense to have this capabilities to just make it easier for people to sign in. But it would probably not pass any type of FIPS or HIPAA certification because the weakest link is the magic email link, leaving the door open to e.g. phishing attacks (on the mail address).

**Comment by aeneasr** [comment]
Interesting Hacker News thread on sign in links: https://news.ycombinator.com/item?id=25628729

**Comment by znerol** [comment]
Interesting doc linked in the HN comments: https://www.w3.org/TR/capability-urls/

I am using one site which does sign-in by email. I think this only works well for low privilege accounts on sites with very long living sessions (months or even years).

**Comment by vincentserpoul** [comment]
just to make it clearer, the email register mentioned is not part of webauthn.
webauthn itself has multiple config, some being more secure than others, but overall, it's quite secure.

**Comment by docelic** [comment]
Could you please clarify - Kratos' description includes "...with ORY-hardened authentication, MFA, FIDO2...".
Does this mean that some of the topics discussed in this thread (for example, TOTP) are implemented, or everything is still work in progress?

**Comment by sycured** [comment]
@docelic actually, I think that it's stale/work in progress: https://github.com/ory/kratos/issues/26#issuecomment-711169742
This answer to my comment for ETA proves that the team is more looking for another thing than implementing everything that we need globally in this style of tool.

For me, the description is just like commercial fraud: we sell something (in the description) that doesn't do the job.

**Comment by aeneasr** [comment]
> For me, the description is just like commercial fraud: we sell something (in the description) that doesn't do the job.

On an open source project that is doing **commercial fraud**? Are you serious?

If you take a look at the current work in several branches and the commit history, instead of making allegations, the question would have been answered already: We are dedicating every day since end of december 2020 on this topic. Maybe have a bit of patience instead of making baseless accusations, or, ask how you can help out. You are free to use a proprietary technology that charges $1 per user per month any time. By the way, if you want to support open source, you can [sponsor our work so that we can develop features faster](https://opencollective.com/ory).

In the future, please follow our[ community guidelines](https://github.com/ory/kratos/blob/master/CONTRIBUTING.md#conduct).

**Comment by nerrixDE** [comment]
@aeneasr fully agree with you there. I was looking forward to contribute, I asked some questions in #kratos in slack for where to get started, an answer there would be much appreciated :heart:

**Comment by aeneasr** [comment]
Thank you @nerrixDE ! It has been quite busy as we have been moving offices (we posted some pictures in the general chat :) ). I am currently refactoring the flows and also the form fields to accustom things such as: QR codes, security back up codes, javascript, and so forth. Unfortunately, that makes it a bit hard to contribute as there are constant changes in the code base!

However, what I think would be helpful, is to better understand how we could implement WebAuth! I played around with it and also bought a YubiKey, but have not yet figured out how to integrate that in the Login UI - given that we need to execute some JavaScript to make it work. Any knowledge you might have there would be helpful!

**Comment by aeneasr** [comment]
By the way - you can follow the progress here: https://github.com/ory/kratos/pull/1012 - I expect the whole effort to take around 2 more months (no guarantees) as we have to rework a lot of the internals to support multi-step flows. We also want to solve password less login with e.g. YubiKeys from the get go, which means that we have to rework a lot of stuff. Because we have so many tests, this takes a ton of time!

**Comment by nerrixDE** [comment]
I really enjoy how @nextcloud implemented WebAuthn. Backend is in PHP but the concept should work for kratos as well.

**Comment by vincentserpoul** [comment]
@aeneasr once you have something stable, let me know, I'm sure I can help quite a bit on the webauthn side

**Comment by Jonas-Sander** [comment]
Kind of random, still just gonna throw this in: If you enroll into the Google Advanced Secruity Program where hardware keys are enforced for authentication you need at least 2 keys so that you have one key as a backup.     
So you don't have any backup codes there.

**Comment by aeneasr** [comment]
Ah nice, that makes sense, thank you for the use case!

**Comment by ashley-3r** [comment]
@aeneasr U2F in Go https://github.com/duo-labs/webauthn. The most secure implementation realistically is password + U2F (with yubikey, titankey, solo https://solokeys.com/), but i would argue that PIN + U2F should be enough. Another U2F can be the backup (when you buy titankeys from Google you get the bluetooth + USB FIDO2). I think recovery keys make sense as well, if Kratos aims to be useful to big organizations as well. I am not a fan of adaptive MFA (it often confuses large scale company users and during real pentests i enjoyed using the user's browser to bypass adpative MFA when they weren't there or use directly their cookies/access tokens when they were present). I would recommend PIN + U2F on 15 mins logout without user actions on browser and using refresh tokens with local reauthentication mobile to unlock keystore. I absolutely abhor it but i was surprised by the number of authentication scenarios in South America, Africa and Asia outside China only using phone identity to authenticate and also by telecom providers in North America upgrading it with Android U2F with ACR would be great.
Authentication methods (no adaptive) :
- WebAuthn (most recent Android support it now and planned iOS)
- Email
- SMS
- TOTP/HOTP
- password with ACR.

**Comment by aeneasr** [comment]
Thank you @ashley-3r for the insights! I think that all makes a lot of sense. Adaptive MFA is also hard to abstract I think as everyone will probably want different rules?

WebAuthn is by the way already supported in iOS! Our plan was to have WebAuthn and TOTP as the first supported authentication factors and also suitable for MFA.

**Comment by nerrixDE** [comment]
yeah but I think it makes sense to have the priority (if prioritizable) to TOTP

**Comment by robertlagrant** [comment]
I don't know if it helps, but I spotted arengu.com on HN, which is a login flow builder. I wonder if ideas from how they structure things might help your thinking in this complex area.

(I have no affiliation with them.)

**Comment by aeneasr** [comment]
That's a nice looking product! Thank you for the tip!

**Comment by dduzgun-security** [comment]
An interesting idea for the MFA flow would be to offer 2 options.

- Classic flow :  `identifier->password->MFA`
- Protected flow : `identifier->MFA->password`

The protected flow would reduce the attack surface of a malicious actor since MFA would be required before entering a password.

**Comment by aeneasr** [comment]
Why would it reduce the attack surface?

**Comment by dduzgun-security** [comment]
**Pros**
Asking for the MFA (that changes over time) before asking for a password would reduce the possibilities of brute force attacks to guess a user's passwords. Knowing that many people tend to reuse passwords, this may be useful. 

**Cons**
However, it would mostly work with TOTP because other methods (Notifications, Email,  SMS, ..) could be used to spam users. Also, it would lead to an enumeration of users that doesn't have the MFA activated.

This method could potentially reduce password brute force attacks for users with TOTP MFA. The more I think about it, I realize that the classical flow with a proper rate-limiting mechanism would make more sense.

**Comment by vincentserpoul** [comment]
Stepping back a little, [MFA means Multi Factor Authentication](https://en.wikipedia.org/wiki/Multi-factor_authentication#Factors).
The user+password is one among other factors of authentication.
The config for authentication could be an selection/combination/order of allowed auths methods, validating one or more factors.

**Comment by vincentserpoul** [comment]
> This makes it way too complicated and doesn't add much if any security. Passwords are a "first" factor, never a second factor. Hardware tokens, TOTP, ... can be used as a first factor if you don't want passwords but not the other way around.

I agree it sounds convoluted, but passwords are in the end just a "what I know" with a (not always) good entropy. Hardware tokens and TOTP are a "what I have" which is usually not enough as a single factor.

That said, I kinda think it'd be error prone if the UX is not carefully crafted.
In any case, in the end it's your decision, just trying to think about a less password centric world :).

**Comment by arianvp** [comment]
I was the one who opened the original ticket regarding FIDO2 here (#158; feel free to close as duplicate )

@aeneasr https://github.com/fido-alliance/how-to-fido/blob/master/HowToFIDO.md is a really nice canonical resource in the most common FIDO workflows that are implemented. I found it really helpful. Indeed e.g. how Azure implements passwordless login is utilizing FIDO2 as a secure reauthentication mechanism but not as a registration mechanism at this point.


For those interested; work has been done on coming up with a key recovery scheme which would make FIDO2 also more suited for primary registration. But I don't think any keys implement it yet https://github.com/Yubico/webauthn-recovery-extension



>  Hardware tokens and TOTP are a "what I have" which is usually not enough as a single factor.

@vincentserpoul note that the hardware tokens can be forced to prompt a pin (and the pin has brute-force protection) so a hardware token is truly  What I have + what I know. 2 factor.

**Comment by nerrixDE** [comment]
Some more / less obvious considerations:

- QR Code shall be generated locally in JS, offline.
- Underlying TOTP master-code should be exposable for manual addition to i.e. a password manager
- Entering the generated 6-digit code mandatory before activation
- Download of 10 backup codes in a .txt file, the file is also generated in browser.
- Those backup codes are invalidated when 2FA is disabled, i.e. a new TOTP-Master is generated.
- No Backup via SMS/Phone for now, this is a) insecure (see simswapping) and requires additional effort not needed for the so-long wanted working 2FA in Kratos.
- Do not add Webauthn or other technologies for now. They can be easily added into an existing MFA stack later, as this is how they were designed.
- Allow to force 2FA with a trait or something settable by the backend, so users can disable 2FA (like when they want to reset it, which is ofc needed when codes were lost or exposed) but they then (if the force-2fa-flag is set on their account) have to re-configure 2FA in order to proceed / login. This is a big deal for Kratos-Users where they need to enforce 2FA because compliance guidelines.
- Store the backup codes hashed, this should be a zero-brainer.
- Keep kratos "bring your own UI" principle. This works great for 2FA as neither QR-Code generation (& display) nor text file generation / download are issues for any UI. Add sample code to kratos sample "SecureApps" but if you are in the fashion to build you own backend around kratos, this won't be the issue.
- Do not seperate backup code usage and normal 2fa code imput fields. While the backup codes should ofc be longer, not only 6 digits numbers, they should be as secure as a generated password, let them be entered in the same HTML field. Not doing so confuses some users and causes additional overhead. Let's keep it simple.
- Define strict anti-bruteforce rules. Sure, 6 digits TOTP codes can be bruteforced, thus protection against that is mandatory. For now let's stick to blocking on a user-basis, i.e. timeouts for that account and maybe require backup code usage when a user-defined threshold of attempts is matched. Further hardening could be integrated with #138.

My personal ideas, maybe some are useful, some are definitely obvious.

**Comment by nerrixDE** [comment]
@aeneasr ^

**Comment by nerrixDE** [comment]
Any possibility we'll see a just-working 2FA in Kratos in Q2 2021?
I am happy to contribute where I can.

**Comment by aeneasr** [comment]
As stated often before we do not give out timelines. As we’re building up SRE resources and onboarding paying customers we do have some shifts in priorities and we’re still very much resource constrained with a small team that now also does SRE.

MFA is still of high priority but so are many other things!

The best way to help speed this up is probably an implementation plan for e.g. TOTP, followed by an implementation. We’re also working with select and interesting companies that can become part of Ory Insider and get access to Ory (the SaaS platform) before it goes GA. Part of Ory Insider is also prioritizing issues - although we can’t promise impossible things :)

Given that it‘s already June, MFA until EoM with all the other business related tasks (e.g. Series A) appears pretty much unrealistic from my side.

**Comment by aeneasr** [comment]
Work on this is currently blocked by: https://github.com/ory/kratos/pull/1367

**Comment by nerrixDE** [comment]
Some ideas (inspired by https://github.com/nextcloud/twofactor_totp), RFC 6238:


Routes (pseudo"code"):


POST /backupcodes/create

- Creates 10 backup codes, stores them (hashed) in kratos database
- Returns those 10 codes, randomly generated, length: 15, a-zA-Z0-9
- Client: Displays them, offers them to be printed or downloaded as .txt (via JS)
- They can be re-generated, however only shown once upon creation obviously


POST /twofactor/enable

- Requires backup codes to be generated before
- Generates ~16 length random secret, a-zA-Z0-9, stores it encrypted (not hashed) in db, overwrites previous. Stores it with flag activated=false
- Returns both the secret and a URL like "otpauth://totp/<SITE NAME>:<USERNAME>@<DOMAIN NAME>?secret=<GENERATED SECRET>&issuer=<SITE NAME>" which JS will turn into a QR-Code to scan

POST /twofactor/activate

- User scanned QR code / optained a 6 digit code, which is then after entered, sent to this endpoint
- Kratos verifies the validity of the onetime-password, and if okay, sets the activated flag = true.


POST /twofactor/disable

- Requires backup code or valid 6 digit onetime-code
- Add global setting to only allow users secret re-generating, not completly disabling 2FA
- Invalidates 2fa master codes, generating new ones as part of the answer

POST /twofactor/validate

- Called as part of the login flow or the twofactor-disable flow
- the posted "code" is either a valid backup code (which is then marked expired in DB) or a 6 digit code, the counter in the DB is increased (this counter is part of the TOTP specification, won't go in the details, but it's actually quite straight-forward)


Neither details of impementating RFC 6238 may be accurately met nor does this consider anti-bruteforce rules among other things.


https://github.com/nextcloud/twofactor_totp and https://github.com/nextcloud/server/tree/master/apps/twofactor_backupcodes are a pretty good resource to look at, their implementation is very easy to reproduce when looking at the code and is pretty well done.

But TOTP is actually quite straight-forward and shouldn't be too hard, would feel comfortable implementing it, if I would be comfortable with Go 🙃

**Comment by nerrixDE** [comment]
Kratos is a great project with much potential, however the always advertised 2FA was constantly moved in the future, I now had deadlines for month-long-projects to match and those were not matched with Kratos, but with a FOSS Alternative in Java. 
Still I see the great potential, especially in zero trust among other Principles, but currently kratos is off the table and I thus probably won't contribute in the coming months. 👋

**Comment by robinmanuelthiel** [comment]
> Kratos is a great project with much potential, however the always advertised 2FA was constantly moved in the future, I now had deadlines for month-long-projects to match and those were not matched with Kratos, but with a FOSS Alternative in Java.
> Still I see the great potential, especially in zero trust among other Principles, but currently kratos is off the table and I thus probably won't contribute in the coming months. 👋

But you do know, that Kratos is Preview Software, right? So relying on certain features for production use cases with deadlines might be your fault, not theirs...

You can read more about the maturity grade of Kratos and other ORY projects here: https://www.ory.sh/docs/ecosystem/versioning/#maturity

**Comment by nerrixDE** [comment]
> So relying on certain features for production use cases with deadlines might be your fault, not theirs..

Of course, I don't blame anyone. I was just hoping that it would take only a few months and that it'd eventually match my dates, however that doesn't matter here.
I can fully understand why Ory is setting priorities as it does, but Ory has to be able to understand that when they are advertising features such as 2FA in the Github Overview which is in "TODO" since 2019(!) that some users are mislead and wrongly consider Kratos for their projects, so did I [and others](https://github.com/ory/kratos/issues/26#issuecomment-766796149) (however I do not share the expectations and tone the user quoted in the reply has torwards this great project).

**Comment by aeneasr** [comment]
v0.7 is out - v0.8 milestone starts now and the only issue assigned to 0.8 is this one 🎉

**Comment by ufukty** [comment]
Glad to see attention. OTP support is a must for many B2Cs. Lack of it would be a dealbreaker.

**Comment by ysaakpr** [comment]
@ufukty  you can do just otp based login in normal user name password  login with qualified mobile number as username and dynamically change the password on sending the otp to customer, and then do password based credential  for logging in.

It should be possible with a little bit UI and dynamic password updates from backend for the user., But yeah, it might break if there is regular email based login also in place, where the credentials are shared. This is a basic missing feature for me as well from kratos.

**Comment by ufukty** [comment]
@ysaakpr Thanks for your answer. I am still trying to understand Ory universe and maybe I got Kratos' place wrong but isn't that redundancy? Why use Kratos if I it will be my job to design and implement basic functionality. I am willing to write an extension/plugin etc for my specific use case if it is supported, but no patch. I am not sure how safe to rely on a hack if we think future updates can break things. I rather native support.

**Comment by aeneasr** [comment]
Something to get you hooked:

<img width="747" alt="Bildschirmfoto 2021-08-05 um 12 49 42" src="https://user-images.githubusercontent.com/3372410/128338258-73b9ad77-c8e1-4464-8448-da8f78a63c6c.png">

Follow the progress in #1624

**Comment by ysaakpr** [comment]
> @ysaakpr Thanks for your answer. I am still trying to understand Ory universe and maybe I got Kratos' place wrong but isn't that redundancy? Why use Kratos if I it will be my job to design and implement basic functionality. I am willing to write an extension/plugin etc for my specific use case if it is supported, but no patch. I am not sure how safe to rely on a hack if we think future updates can break things. I rather native support.

Yeah,  there will be potential risk involved in using a system, which supposed to be used in its defined way. But that risk is what you need to understand mitigate by evaluation and regular updates probably till the required feature comes.  **I think this request should be raised on a different issue altogether**, because this is mainly focusing on 2FA/MFA, **mobile OTP login is not 2FA, its yet another login mechanism as similar to email password.**

**Comment by ufukty** [comment]
@ysaakpr Then we didn’t understand each other from start. I don’t want to use OTP solely. I was thinking OTP is the second part of MFA. But you use specifically TOTP term for that usage as I understand. Thanks tho

**Comment by aeneasr** [comment]
Linking & unlinking works (no tests yet). Next up will be a PoC login flow

https://user-images.githubusercontent.com/3372410/128348645-68f36844-16a1-472e-add3-640905788b63.mov

**Comment by aeneasr** [comment]
@ufukty @ysaakpr thank you for your ideas - please keep this on topic though. This issue is about implementing TOTP, WebAuth, and recovery codes in Ory Kratos. What you want is probably "passwordless" login which is a different mechanism and it will be possible to implement that easily once https://github.com/ory/kratos/issues/1451 lands.

We have a wonderful [slack community](https://slack.ory.sh) where you can continue the conversation if you'd like - thank you! :)

**Comment by aeneasr** [comment]
So one important question is how the login should work. Basically, we have two options:

1. Require second factor to receive a session
2. Do not require second factor for receiving a session, but add the authentication level to the session

## We enforce MFA

The first option would mean that every user with MFA set up would need to complete the 2FA challenge in order to receive a session. This in turn implies:

* Unless we provide a configuration option, the user must always authenticate using 2FA to get a session
* If 2FA is required site-wide it would imply that the user needs to set up MFA during registration, or by implementing a configurable "grace" period
* If we provide a config option it means we need to differentiate between a session with 2fa and a session without 2fa

## Application enforces MFA

Alternatively, we could not require 2fa to receive a session but instead add information to the session with factor was authenticated when - for example:

```
{
  "id": "...",
  "authenticated_at": "<some-later-date>",
  "authenticator": {
    "assurance_level": "aal2",
    "methods": [
      {
         "method": "password",
         "authenticated_at": "<some-date>"
      },
      {
         "method": "totp",
         "authenticated_at": "<some-later-date>"
      },
    ]
  }
}
```

AAL refers to 4.2 Authenticator Assurance Level: https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-63b.pdf

Then, the application's middleware / proxy has to decide what AAL is acceptable. For example, it might be acceptable to use AAL1 for some components, but it is required to have AAL2 for others. If AAL2 is not reached, the middleware could redirect to Ory Kratos' login endpoint requesting elevation of the AAL.

We theoretically could implement a hook which redirects to the MFA UI immediately after login - however, it would still issue a session! So you would still need to implement the AAL check in the middleware.

## Conclusion

In conclusion, I believe that the second option is a bit more work on the adopter's side but it also comes with the highest degree of freedom.

It would be easily possible to implement a check to see what the AAL is and redirect to the MFA login if it is not sufficient.

Implementing the enforcement in Ory Kratos would mean quite a lot of configuration options and UX questions:

1. How do we set up MFA during registration? It is usually a multi-step process with recovery codes, qr codes, verification, ...
2. How do we enforce MFA for some, and make MFA optional for others? This can refer both to end-users (e.g. only users from group B need MFA) as well as globally (e.g. all users need / do not need MFA)
3. How are we dealing with end-users which remove MFA devices from their account if it is enforced? Do we block it? Can they lock themselves out?

I believe there are more related questions but I think it is obvious where this is going. Implementing MFA enforcement in Ory Kratos would require some type of policy engine which has quite a few paths in terms of user journey. Depending on the implementor, this could get arbitrarily complex.

So while the second option is the "lazier" one for us, as we don't have to implement a policy engine, it also gives more freedom and clarity. Want MFA enforcement? Redirect in the middleware. Don't want it? Don't redirect. Only for some user groups? Solve it in the middleware - and so on.

I also think that it is in the spirit of Ory Kratos and Ory in general which tries to have clear domain boundaries. So Ory Kratos is responsible for authenticating the user and telling your system what we authenticated. The system then decides if it is acceptable or not.

**Comment by schreddies** [comment]
From security perspective it would be more secure that user cannot get any session without 2fa - if attacker conduct successful attack on user credentials and the session will be created, it might give attacker some data through endpoints for authenticated user without 2fa (and leaving the decision engine on the 'client' side might also result in lowers security level).
However kratos gets more complicated in terms of forcing, when forcing, and other complicated rules and requirement for 2fa.

**Comment by aeneasr** [comment]
I agree, if the authentication level is not properly checked in the middleware, there is the possibility that an attacker might gain access to privileged information.

Additionally, it is correct that the `/sessions/whoami` endpoint would return PII as the session is authenticated. So it would be possible to extract information about the user after a phishing attack, even if the application itself can not be accessed.

It could also be possible to access the identity settings page as we would (at least so far) not have an MFA enforcement in place there, except for when a privileged factor is changed.

To address this, here are some ideas:

1. Unless explicitly allowed, `/sessions/whoami` returns an error message indicating that AAL2 is required. This would prevent leakage of PII on this endpoint as we need AAL2 to access it.
2. Unless explicitly allowed, `/settings` would also require AAL2 if it is detected that the user has a second factor configured.

This could be done with a setting called e.g. "soft":

```
selfservice:
  soft_aal2: true
```

What do you think?

**Comment by dadrus** [comment]
To be honest I don't understand the first option. Do you mean the session is only created, when the second factor has been verified, but not, when the first one is verified, or do you mean issuing a specific session object for each factor? Specific session object for a factor sounds strange to me. So I assume you mean, the session is only created, when all required factors are authenticated.
In this case we have actually take into account the needs of the application. If the application has a strong requirement of having two factor authenticated, we should be able to configure that. In this case I would expect no session is issued as long as all required factors are not verified. If however step up authentication is required by the application, so that accessing specific parts of it is possible having only one factor verified and accessing yet another part is only possible when two or more factors are verified, we should be able to support that as well. In this case a session should be issued when just one factor is authenticated and this session object is updated or superseded by another one, when another factor has been used by the user.

**Comment by schreddies** [comment]
@aeneasr Yes, It could be done with some generic mechanism - abstraction for tenant configuration where explicitly client could setup which endpoints (or groups of endpoints - ie. critical ones or PII) could be access without 2fa. Also for future endpoints could be filled with annotation for that purpose. 
@dadrus same concept, from what I can get!

**Comment by aeneasr** [comment]
@dadrus basically the difference is if we want non-step up auth enforced in Ory Kratos or not and if we do, how to implement it. Enforcement can become extremely complex (see KeyCloak docs: https://github.com/keycloak/keycloak-community/blob/master/design/multi-factor-admin-and-step-up.md) and we would really like to avoid it.

**Comment by dadrus** [comment]
Ok. Got it. Yeah, KeyCloak is very complex in this regards. 

I'm with you and @schreddies on how it could be achieved. I think having support of something like "configurable annotations" for kratos endpoints which cannot be accessed without mfa for a specific tenant combined with the freedom for the application to enforce mfa or to have step up authentication is the way to go.

**Comment by aeneasr** [comment]
Cool, awesome to have you on board with this one!

**Comment by schreddies** [comment]
@aeneasr The other question is: could be use the OPA (Open Policy Agent) for making such rules as mentioned in the Keycloack? It seems to make sense to store attributes for authentication when, where and how (different methods) 2FA should be forced. That would be the most flexible but also kind of hard to setup for users. And it would be @james-caresnap solution - I guess. 
Side note: in case of having 'annotations' or other check mechanism (authZ check with oathkeeper ie.) the 'blank' could be issued for the user without compromising security

**Comment by aeneasr** [comment]
@schreddies while I like interop between services, I would also like to avoid dependencies on third party systems. In my experience, I found it very difficult to learn and write correct rego and even more so efficient rego. I think simplicity wins here! :)

@james-caresnap - thank you for the example. After sleeping over it and also reading the comments here it seems like this is really the way to go!

The session struct has (on my local fork) already an AAL (Authenticator Assurance Level) which depends on the AMR (Authentication Method References). Both of which are included in the session struct.

The idea would be to specify the required AAL for the different endpoints - so (example):

```
selfservice:
  settings:
    minimum_aal: aal2 # would require first + second factor before being able to access this endpoint

session:
  whoami_minimum_aal: aal2 # would require first + second factor before being able to access the sessions endpoint
```

**Comment by aeneasr** [comment]
Based on the input and ideas from above, I have implemented a login flow. It works really well! There are of course still a couple of things missing (including not having any tests so far) and some bugs but we are getting there on a high level :)

https://user-images.githubusercontent.com/3372410/129038614-c418c709-04d3-4b9d-88ef-e31d18b4cb09.mp4

**Comment by schreddies** [comment]
That's cool, let us see the code! 🔥

**Comment by aeneasr** [comment]
I usually sync my changes every couple of hours with upstream, so you can find them in the PR :)

**Comment by james-caresnap** [comment]
On the very last frame of your video it shows two `totp` authentications separated by fractions of a second. Bug?

**Comment by aeneasr** [comment]
> and some bugs but we are getting there on a high level :)

;)

**Comment by aeneasr** [comment]
MFA is now (as far as I can tell) bugfree and fully integration tested. Next, we need lookup secrets. These are codes (e.g. 8 or 10 chars) which you can use as a back up ONCE when you lose your device. Usually, one gets like 6-12 such lookup codes.

One issue with this is - what do we do if the user runs out of such codes? I assume most platforms would redirect you to a screen where you see: "oh these are my back up codes I need to download them now" before being able to continue with the platform. I will try out how e.g. google and others behave to get a better picture of it.

But yeah, I think that's what's needed here. So I am wondering how on earth we could implement this 😅 One possibility would be to "abuse" the login screen for this. If a look up secret was used, and it was the last look up secret, before you are able to continue, you see the new look up secrets.

**Comment by aeneasr** [comment]
Ok, that's interesting - if you use up all of your backup codes in Google then - well - you have used up all your back up codes. There is no dialog, regeneration, or anything else which tells you: you ran out of back up codes. So basically, if you mess it up, you are locked out of your account.

I think my proposal which would work like this:

1. Request MFA login
2. Lost TOTP -> use back up code, back up code was last code used
3. Show new back up codes and "confirm" button
4. Upgrade session to MFA session and use back up code ONLY after user hits "confirm"

would be much safer!

**Comment by aeneasr** [comment]
So configuring look up secrets now works:

https://user-images.githubusercontent.com/3372410/129366015-189a3441-94e6-47fa-a43a-ac496e70d073.mov

Regarding https://github.com/ory/kratos/issues/26#issuecomment-898366286 - I think this is generally a good idea, but it's also pretty difficult to implement. For now I think it is enough if we require the user - as does Google - to do this on their own. I have tracked this idea in a separate issue: https://github.com/ory/kratos/issues/1649

**Comment by aeneasr** [comment]
Aaaand we got login:

https://user-images.githubusercontent.com/3372410/129369344-7e4eb4f1-25e3-4266-918c-c7e64c9c74cf.mp4

**Comment by aeneasr** [comment]
Well, I kinda got bored and thought why not give WebAuthn a PoC try?

https://user-images.githubusercontent.com/3372410/129558836-34a04243-9917-4dd8-af94-158588e79c81.mov

Just wanted to note on this occasion also that all of this was only possible because we refactored the complete UI and flow types internally. It so powerful and flexible now that even adding webauth can happen in a few hours (without tests of course).

**Comment by twnk** [comment]
> Ok, that's interesting - if you use up all of your backup codes in Google then - well - you have used up all your back up codes. There is no dialog, regeneration, or anything else which tells you: you ran out of back up codes. So basically, if you mess it up, you are locked out of your account.
> ...
>     2. Lost TOTP -> use back up code, back up code was last code used

Just my opinion, but I think it would be much more user-friendly to remind users after each use of a backup code "You have n backup codes remaining, do you want to generate new codes?" Maybe with a different strength of reminder as you get close to 0 remaining. E.g. Attempt to force them to renew if 3 or below.

**Comment by aeneasr** [comment]
WebAuthn MFA login...

https://user-images.githubusercontent.com/3372410/129592759-a40282eb-abcf-4885-b499-65174cc3fbd0.mov

**Comment by aeneasr** [comment]
Yo, it's end-to-end tested now.

https://user-images.githubusercontent.com/3372410/129912849-8c530d90-d011-45d9-9822-7f1afbe349c0.mov


https://user-images.githubusercontent.com/3372410/129912871-5d9e4c60-91f4-4c22-8748-eae6f914c2e2.mov


https://user-images.githubusercontent.com/3372410/129912823-e6d0310b-beed-4edc-83c7-383a9e4bd49d.mov

**Comment by aeneasr** [comment]
TOTP and recovery flows now also work with mobile (webauthn only targets browsers). So basically this is all done except for docs. Hence, unpinning!

Please be patient with us while we release this. Still needs docs and stuff. But should land latest end of september. In the meanwhile, you'll be able to run this off of master :)

**Comment by arianvp** [comment]
Awesome! 

>  (webauthn only targets browsers)

What exactly do you mean by this remark? It'll still work on a mobile web browser right (E.g. iPhone exposes TouchID as a webauthn platform authenticator).  As I'd say mobile authentication is actually the prime usecase of webauthn. (Most people have platform authenticators; but not yubikeys)

**Comment by aeneasr** [comment]
WebAuthn only works on browsers. If you build a browser app, and use the mobile browser, then it's not a native app. A native app is Swift / Java. There, WebAuthn does not work, because it is a W3C API.

**Comment by schreddies** [comment]
About this comment @aeneasr 

>  Lost TOTP -> use back up code, back up code was last code used
We are talking about last used code for successful authentication? It's the place where some brute-forcing should be consider as potential attack vector - it will require some protection against it. TBH. All that kind of recover functions, somehow lowering security of prime mechanism should implement that. 

About OPA and rego. Yes, it's painful :D

**Comment by lsancho** [comment]
there is a expected date to complete this issue?
